SYCL/InvokeSimd/Regression/nbarrier_multiple.cpp

```diff
@@ -0,0 +1,114 @@
+// TODO: enable on Windows once driver is ready
+// NOTE: named barrier supported only since GEN12
+// REQUIRES: gpu-intel-gen12 && linux
+// UNSUPPORTED: cuda || hip
+//
+// TODO: enable when Jira issue resolved
+// XFAIL: gpu-intel-gen12
+//
+// RUN: %clangxx -fsycl -fno-sycl-device-code-split-esimd -Xclang -fsycl-allow-func-ptr -std=c++20 -DXETPP_NEW_XMAIN -DUSE_OPENBLAS %s -o %t.out
+// RUN: env IGC_VCSaveStackCallLinkage=1 IGC_VCDirectCallsOnly=1 %GPU_RUN_PLACEHOLDER %t.out
+//
+// VISALTO enable run
+// RUN: env IGC_VISALTO=63 IGC_VCSaveStackCallLinkage=1 IGC_VCDirectCallsOnly=1 %GPU_RUN_PLACEHOLDER %t.out
+
+/*
+ * Test check support for named barriers in invoke_simd.
+ * This case contains two such barriers.
+ */
+
+#define XETPP_CODE_BASE __ESIMD__
+#define ESIMD_XE_HPC
+
+#include <sycl/ext/intel/esimd.hpp>
+#include <sycl/ext/oneapi/experimental/invoke_simd.hpp>
+#include <sycl/sycl.hpp>
+
+using namespace sycl;
+using namespace sycl::ext::oneapi::experimental;
+namespace esimd = sycl::ext::intel::esimd;
+
+#include <sycl/ext/intel/esimd.hpp>
+#include <sycl/ext/intel/experimental/esimd/memory.hpp>
+#include <sycl/ext/oneapi/experimental/invoke_simd.hpp>
+#include <sycl/sycl.hpp>
+
+#include <functional>
+#include <iostream>
+#include <type_traits>
+
+/* Subgroup size attribute is optional
+ * In case it is absent compiler decides what subgroup size to use
+ */
+#ifdef IMPL_SUBGROUP
+#define SUBGROUP_ATTR
+#else
+#define SUBGROUP_ATTR [[intel::reqd_sub_group_size(VL)]]
+#endif
+
+using namespace sycl::ext::oneapi::experimental;
+namespace esimd = sycl::ext::intel::esimd;
+
+constexpr int Size = 4096;
+constexpr int VL = 16;
+using namespace sycl;
+
+class Test {};
+
+__attribute__((always_inline)) void
+named_barried_run(nd_item<1> *ndi) SYCL_ESIMD_FUNCTION {
+  const uint8_t BARNUM = 32;
+  sycl::ext::intel::experimental::esimd::named_barrier_init<BARNUM>();
+
+  uint8_t barrier_id = 1;
+  uint8_t producer_consumer_mode = 0;
```

Comment:
i might be reading the header wrong but is 0 valid? my reading was only 0x1 or 0x2 was valid

Reply:
I believe 0 value stands for producer-consumer mode. In this case producer thread waits for the completion notification to synchronize with all the other threads.
At least that's how I used this mode in ESIMD named barriers tests: https://github.com/intel/llvm-test-suite/blob/intel/SYCL/ESIMD/named_barriers/exec_in_order.cpp

Reply:
ah ok thanks for the explanation